sources/api/netdog/src/dns.rs

```diff
@@ -0,0 +1,259 @@
+//! The dns module contains the code necessary to gather DNS settings from config file, DHCP lease
+//! file, or an already-parsed DHCP lease via LeaseInfo struct.  It also contains the code
+//! necessary to write a properly formatted `resolv.conf`.
+use crate::lease::LeaseInfo;
+use crate::RESOLV_CONF;
+use rand::prelude::SliceRandom;
+use rand::thread_rng;
+use serde::Deserialize;
+use snafu::ResultExt;
+use std::collections::BTreeSet;
+use std::fmt::Write;
+use std::fs;
+use std::iter::FromIterator;
+use std::net::IpAddr;
+use std::path::Path;
+
+static DNS_CONFIG: &str = "/etc/netdog.toml";
+
+#[derive(PartialEq, Debug, Deserialize)]
+pub(crate) struct DnsSettings {
+    #[serde(rename = "name-servers")]
+    nameservers: BTreeSet<IpAddr>,
+    #[serde(rename = "search-list")]
+    search: Option<Vec<String>>,
+}
+
+impl DnsSettings {
+    /// Create a DnsSettings from TOML config file
+    pub(crate) fn from_config() -> Result<Option<Self>> {
+        Self::from_config_impl(DNS_CONFIG)
+    }
+
+    fn from_config_impl<P>(path: P) -> Result<Option<Self>>
+    where
+        P: AsRef<Path>,
+    {
+        // Ensure we don't attempt to load a nonexistent or empty config file At boot time, the
+        // config file won't exist because it hasn't been generated by the API yet.  After that,
+        // the file will always exist because it's a configuration file for settings in the API and
+        // will always be generated.  If the DNS settings aren't populated, the file will be empty.
+        // We can assume if the file is empty that the settings don't exist.
+        let path = path.as_ref();
+        let config_exists = if Path::exists(path) {
+            let file_len = fs::metadata(path)
+                .context(error::DnsConfMetaSnafu { path })?
+                .len();
+            file_len != 0
+        } else {
+            false
+        };
```

Comment:
Nit: How about this:

```rust
let config_exists = Path::exists(path) && fs::metadata(path)
    .context(error::DnsConfMetaSnafu { path })?
    .len() != 0
```

Reply:
I mention this is the (rather lengthy) comment above this code but the tl;dr is that on first boot this file won't exist (which is fine) and I don't want to return an error when attempting to run `fs::metadata()` against it.

I initially wrote some code running `fs::metadata()` and matching on the error that was returned, but I thought that code was less readable and clear than this.  I'm totally willing to debate that and change it though. :)